processing/src/main/java/io/druid/query/groupby/epinephelinae/ConcurrentGrouper.java

```diff
@@ -292,19 +325,413 @@ private boolean isParallelSortAvailable()
                    .collect(Collectors.toList());
   }
 
+  /**
+   * Merge dictionaries of {@link Grouper.KeySerde}s of {@link Grouper}s.  The result dictionary contains unique string
+   * string keys.
+   *
+   * @return merged dictionary if its size does not exceed max dictionary size.  Otherwise null.
+   */
+  @Nullable
+  private List<String> mergeDictionary()
+  {
+    final long maxDictionarySize = combineKeySerdeFactory.getMaxDictionarySize();
+    final Set<String> mergedDictionary = new HashSet<>();
+    long totalDictionarySize = 0L;
+
+    for (SpillingGrouper<KeyType> grouper : groupers) {
+      final List<String> dictionary = grouper.getDictionary();
+      totalDictionarySize += dictionary.stream().mapToLong(RowBasedGrouperHelper::estimateStringKeySize).sum();
```

Comment:
This algorithm may conclude that totalDictionarySize exceeds the maximum, but actually it's not: it accounts keys that may already be in the merged dictionary.

Reply:
Good point. Fixed.